pkg/api/validate/openshiftcluster_validatedynamic.go

```diff
@@ -59,66 +60,162 @@ func (dv *openShiftClusterDynamicValidator) Dynamic(ctx context.Context) error {
 	if err != nil {
-		return err
+		if _, isCloudError := err.(*api.CloudError); isCloudError {
+			return err
+		}
+		dv.log.Error(err)
+		return api.NewCloudError(
+			http.StatusBadRequest,
+			api.CloudErrorCodeInvalidParameter,
+			"", "Failed to retrieve an AAD token for the service principal",
+		)
 	}
 
 	spAuthorizer := refreshable.NewAuthorizer(token)
 
 	spDynamic, err := dynamic.NewValidator(dv.log, dv.env, dv.env.Environment(), dv.subscriptionDoc.ID, spAuthorizer, dynamic.AuthorizerClusterServicePrincipal)
 	if err != nil {
-		return err
+		if _, isCloudError := err.(*api.CloudError); isCloudError {
+			return err
+		}
+		dv.log.Error(err)
+		return api.NewCloudError(
+			http.StatusBadRequest,
+			api.CloudErrorCodeInvalidParameter,
+			"", "Failed to validate the service principal token",
+		)
 	}
 
 	// SP validation
 	err = spDynamic.ValidateServicePrincipal(ctx, spp.ClientID, string(spp.ClientSecret), dv.subscriptionDoc.Subscription.Properties.TenantID)
 	if err != nil {
-		return err
+		if _, isCloudError := err.(*api.CloudError); isCloudError {
+			return err
+		}
+		dv.log.Error(err)
+		return api.NewCloudError(
+			http.StatusBadRequest,
+			api.CloudErrorCodeInvalidParameter,
+			"", "Failed to validate the cluster service principal",
+		)
 	}
 
 	err = spDynamic.ValidateVnet(ctx, dv.oc.Location, subnets, dv.oc.Properties.NetworkProfile.PodCIDR, dv.oc.Properties.NetworkProfile.ServiceCIDR)
 	if err != nil {
-		return err
+		if _, isCloudError := err.(*api.CloudError); isCloudError {
+			return err
+		}
+		dv.log.Error(err)
+		return api.NewCloudError(
+			http.StatusBadRequest,
+			api.CloudErrorCodeInvalidLinkedVNet,
+			"", "Invalid cluster virtual network CIDRs",
+		)
 	}
 
 	err = spDynamic.ValidateSubnets(ctx, dv.oc, subnets)
 	if err != nil {
-		return err
+		if _, isCloudError := err.(*api.CloudError); isCloudError {
+			return err
+		}
+		dv.log.Error(err)
+		return api.NewCloudError(
+			http.StatusBadRequest,
+			api.CloudErrorCodeInvalidParameter,
+			"", "Invalid cluster virtual network subnets",
+		)
 	}
 
 	err = spDynamic.ValidateProviders(ctx)
 	if err != nil {
-		return err
+		if _, isCloudError := err.(*api.CloudError); isCloudError {
+			return err
+		}
+		dv.log.Error(err)
+		return api.NewCloudError(
+			http.StatusBadRequest,
+			api.CloudErrorCodeInvalidParameter,
+			"", "Failed to validate the registered resource providers",
+		)
 	}
 
 	err = spDynamic.ValidateQuota(ctx, dv.oc)
 	if err != nil {
-		return err
+		if _, isCloudError := err.(*api.CloudError); isCloudError {
+			return err
+		}
+		dv.log.Error(err)
+		return api.NewCloudError(
+			http.StatusBadRequest,
+			api.CloudErrorCodeQuotaExceeded,
+			"", "Current resource quotas do not permit the creation of the cluster",
```

Comment:
This too is not always correct. We can fail compute usage, for example:

https://github.com/Azure/ARO-RP/blob/4a48003b3e2345fda51ac3e860df4134cb494158/pkg/api/validate/dynamic/quota.go#L126-L133

Or network usage:

https://github.com/Azure/ARO-RP/blob/4a48003b3e2345fda51ac3e860df4134cb494158/pkg/api/validate/dynamic/quota.go#L142-L145

It doesn't mean that that quatas do not permit cluster. It means that something went wrong internally and it might not be actionable by the customer.